src/_avif.c

```diff
@@ -0,0 +1,819 @@
+            avifResultToString(result));
+        avifDecoderDestroy(self->decoder);
+        self->decoder = NULL;
+        Py_DECREF(self);
+        return NULL;
+    }
+
+    if (self->decoder->alphaPresent) {
+        self->mode = "RGBA";
+    } else {
+        self->mode = "RGB";
+    }
+
+    return (PyObject *)self;
+}
+
+PyObject *
+_decoder_dealloc(AvifDecoderObject *self) {
+    if (self->decoder) {
+        avifDecoderDestroy(self->decoder);
+    }
+    PyMem_Free(self->data);
+    Py_RETURN_NONE;
+}
+
+PyObject *
+_decoder_get_info(AvifDecoderObject *self) {
+    avifDecoder *decoder = self->decoder;
+    avifImage *image = decoder->image;
+
+    PyObject *icc = NULL;
+    PyObject *exif = NULL;
+    PyObject *xmp = NULL;
+    PyObject *ret = NULL;
+
+    if (image->xmp.size) {
+        xmp = PyBytes_FromStringAndSize((const char *)image->xmp.data, image->xmp.size);
+    }
+
+    if (image->exif.size) {
+        exif =
+            PyBytes_FromStringAndSize((const char *)image->exif.data, image->exif.size);
+    }
+
+    if (image->icc.size) {
+        icc = PyBytes_FromStringAndSize((const char *)image->icc.data, image->icc.size);
+    }
+
+    ret = Py_BuildValue(
+        "IIIsSSS",
+        image->width,
+        image->height,
+        decoder->imageCount,
+        self->mode,
+        NULL == icc ? Py_None : icc,
+        NULL == exif ? Py_None : exif,
+        NULL == xmp ? Py_None : xmp);
+
+    Py_XDECREF(xmp);
+    Py_XDECREF(exif);
+    Py_XDECREF(icc);
+
+    return ret;
+}
+
+PyObject *
+_decoder_get_frame(AvifDecoderObject *self, PyObject *args) {
+    PyObject *bytes;
+    PyObject *ret;
+    Py_ssize_t size;
+    avifResult result;
+    avifRGBImage rgb;
+    avifDecoder *decoder;
+    avifImage *image;
+    uint32_t frame_index;
+
+    decoder = self->decoder;
+
+    if (!PyArg_ParseTuple(args, "I", &frame_index)) {
+        return NULL;
+    }
+
+    result = avifDecoderNthImage(decoder, frame_index);
+
+    if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Failed to decode frame %u: %s",
+            decoder->imageIndex + 1,
+            avifResultToString(result));
+        return NULL;
+    }
+
+    image = decoder->image;
+
+    memset(&rgb, 0, sizeof(rgb));
+    avifRGBImageSetDefaults(&rgb, image);
+
+    rgb.depth = 8;
+
+    if (decoder->alphaPresent) {
+        rgb.format = AVIF_RGB_FORMAT_RGBA;
+    } else {
+        rgb.format = AVIF_RGB_FORMAT_RGB;
+        rgb.ignoreAlpha = AVIF_TRUE;
+    }
+
+    avifRGBImageAllocatePixels(&rgb);
+
+    Py_BEGIN_ALLOW_THREADS
+    result = avifImageYUVToRGB(image, &rgb);
+    Py_END_ALLOW_THREADS
+
+    if (result != AVIF_RESULT_OK) {
+        PyErr_Format(
+            exc_type_for_avif_result(result),
+            "Conversion from YUV failed: %s",
+            avifResultToString(result));
+        avifRGBImageFreePixels(&rgb);
+        return NULL;
+    }
+
+    size = rgb.rowBytes * rgb.height;
```

Comment:
Is this guaranteed to not overflow, even in the face of invalid input?

Reply:
libavif currently restricts images to a maximum of 2^28 pixels. If the dimensions are larger than 16384x16384 then the function that sets `decoder->image->width` and `decoder->image->height` fails. So I suppose that a 4-channel 16384x16384 8-bit image could overflow on a 32-bit platform. I'm not certain because the codecs used by libavif have their own overflow limit checks. For instance, dav1d enforces a maximum of 2^26 pixels on 32-bit systems. Should I add a check against `PY_SSIZE_T_MAX` to be sure?

Reply:
Added here

https://github.com/python-pillow/Pillow/blob/b84a8e0d47d9e8807d207aba9c9eeb455646ef62/src/_avif.c#L619-L622

Reply:
Basically, I'm the one who will get a CVE on this if there's a problem, and I'd like really clear guidelines about what the assumptions are for sizes of things and where they come from for dangerous operations like memset, malloc, and pointer reads/writes.  This isn't so much for now, but a couple years down the line, things need to be clear. This will be fuzzed, this will be run under valgrind, so hopefully there won't be problems. 

I've basically had to reverse engineer how SgiRleDecode works over the last month or so, and I'd like to be preventing that sort of experience in the future.

Reply:
Does raising a `MemoryError` if `rgb.height > PY_SSIZE_T_MAX / row_bytes` (as I have in the latest PR push) suffice to address that concern?